src/main/scala/pirate.internal/ParseTraversal.scala

```diff
@@ -176,17 +194,45 @@ object ParseTraversal {
         })
     }
 
-  def eval[A](p: Parse[A]): Option[A] = p match {
+  def parseError[A](arg: String): P[A] =
+    arg.toList match {
+      case '-' :: _ => errorMessageP(s"Invalid option `${arg}`")
+      case _        => errorMessageP(s"Invalid argument `${arg}`")
+    }
+
+  def eval[A](p: Parse[A]): ValidationNel[MissingParser, A] = p match {
     case ValueParse(o) =>
-      o
+      o.successNel
     case ParserParse(p) =>
-      None
+      p match {
+        case SwitchParser(meta, _) =>
+          MissingFlag(meta.names.get.render).failureNel
+        case FlagParser(meta, _) =>
+          MissingFlag(meta.names.get.render).failureNel
+        case ArgumentParser(meta, p) =>
+          MissingArg(meta.metavar).failureNel
+        case CommandParser(sub) =>
+          MissingCommand(sub.name).failureNel
```

Comment:
I would be tempted to pass the data down, and resolve the name/meta at the last second. You'll have more options at display time then, plus stronger types.